qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -441,188 +444,224 @@ def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
-                    f"x_vec has {x_vec.shape[1]} dimensions, "
-                    f"but feature map has {self._feature_map.num_parameters}."
-                ) from AttributeError
+                    "x_vec and y_vec have incompatible dimensions.\n"
+                    f"x_vec has {x_vec.shape[1]} dimensions, but y_vec has {y_vec.shape[1]}."
+                )
 
-        if y_vec is not None and y_vec.shape[1] != self._feature_map.num_parameters:
-            raise ValueError(
-                "y_vec and class feature map have incompatible dimensions.\n"
-                f"y_vec has {y_vec.shape[1]} dimensions, but feature map "
-                f"has {self._feature_map.num_parameters}."
-            )
+            self._validate_length(y_vec[0, :], "y_vec")
 
-        # determine if calculating self inner product
-        is_symmetric = True
-        if y_vec is None:
-            y_vec = x_vec
-        elif not np.array_equal(x_vec, y_vec):
-            is_symmetric = False
+        return x_vec, y_vec
 
-        # initialize kernel matrix
-        kernel = np.zeros((x_vec.shape[0], y_vec.shape[0]))
+    def _validate_length(
+        self, vec: ParameterVector | np.ndarray, vec_name: str, adjust_feature_map: bool = False
+    ) -> None:
+        """
+        Validates the size (length) of the vector against the number of parameters of the feature
+        map. Updates the feature map if required to the size of the vector.
 
-        # set diagonal to 1 if symmetric
-        if is_symmetric:
-            np.fill_diagonal(kernel, 1)
+        Args:
+            vec: A vector to validate.
+            vec_name: A vector name, if an error is raised it ise used in the error message.
+            adjust_feature_map: If ``True`` tried to adjust the number of qubits of the feature map.
 
+        Raises:
+            ValueError: If the size of the vector is not the same as the number of parameters of the
+                feature map or when the feature map does not allow to change the number of qubits.
+        """
+        if vec is not None and len(vec) != self._feature_map.num_parameters:
+            raise_error = True
+            if adjust_feature_map:
+                try:
+                    # an attempt to update the feature map with the required number of qubits
+                    self._feature_map.num_qubits = vec
+                    raise_error = False
+                except AttributeError:
+                    pass
+
+            if raise_error:
+                raise ValueError(
+                    f"{vec_name} and class feature map have incompatible dimensions.\n"
+                    f"{vec_name} has {len(vec)} dimensions, "
+                    f"but feature map has {self._feature_map.num_parameters}."
+                )
+
+    def _calculate_kernel_statevector(
+        self, x_vec: np.ndarray, y_vec: np.ndarray, is_symmetric: bool, kernel: np.ndarray
+    ):
```

Comment:
Return type hint missing. Please check other places.

Reply:
Reviewed type hints here and there.